content/blog/2019-11-13-pre/index.md

```diff
@@ -0,0 +1,90 @@
++++
+title = "Partial Redundancy Elimination using Lazy Code Motion for LLVM"
+extra.author = "Siqiu Yao"
+extra.bio = """
+  [Siqiu Yao](http://www.cs.cornell.edu/~yaosiqiu/) is a 3rd year Ph.D. student interested in security about systems and programming languages.
+"""
+
++++
+
+### Problem
+This project aims to write a pass for LLVM that implements *partial redundancy elimination* (PRE) using *[lazy code motion](https://dl.acm.org/citation.cfm?id=143136)* (lcm).
+
+### Algorithm Design
+The algorithm is the same as the one I used in my project 2 [BrilPRE](https://www.cs.cornell.edu/courses/cs6120/2019fa/blog/brilpre/).
+
+
+### Implementation
+This project is implemented as an LLVM pass, and it is based on [llvm-pass-skeleton](https://github.com/sampsyo/llvm-pass-skeleton). 
+The full implementation repository is [here](https://github.com/Neroysq/llvm-pre).
+
+Although the algorithm is the same, LLVM IR is way more complex than Bril, and its framework also takes more time to learn than building from scratch when I dealt with Bril. Here I'd like to talk about the biggest challenge and also the most significant difference from BrilPRE: LLVM IR's SSA form.
+
+In the LLVM IR framework, 
+There are no names for local variables/registers (at least in compiler-generated code). 
+There are only pointers representing the results of instructions, 
+which means I cannot create a new variable with a specific name. 
+
+I can use built-in functions to create an instruction and get the result pointer, 
+but the real trouble is that I don't find a way to implement this: 
```

Comment:
More pertinently: "There is no way to do this:"

It's not just that you couldn't find a way to do this; it's actually impossible because of the SSA form.